File: sdk/python/apibrew/model/resource_action.py
from datetime import datetime
from typing import List, Optional
from enum import Enum
from apibrew.entity import Entity, EntityInfo
from __future__ import annotations

from apibrew.model.resource import Resource



class SubType:
    name: str
    title: str
    description: str
    properties: list[Property]

class Reference:
    resource: Resource
    cascade: bool
    backReference: str

class Property:
    name: str
    type: Type
    typeRef: str
    primary: bool
    required: bool
    unique: bool
    immutable: bool
    length: int
    item: Property
    reference: Reference
    defaultValue: dict
    enumValues: list[str]
    exampleValue: dict
    title: str
    description: str
    annotations: dict[str, str]

class AuditData:
    createdBy: str
    updatedBy: str
    createdOn: datetime
    updatedOn: datetime


class Type(Enum):
    BOOL = "BOOL"
    STRING = "STRING"
    FLOAT32 = "FLOAT32"
    FLOAT64 = "FLOAT64"
    INT32 = "INT32"
    INT64 = "INT64"
    BYTES = "BYTES"
    UUID = "UUID"
    DATE = "DATE"
    TIME = "TIME"
    TIMESTAMP = "TIMESTAMP"
    OBJECT = "OBJECT"
    MAP = "MAP"
    LIST = "LIST"
    REFERENCE = "REFERENCE"
    ENUM = "ENUM"
    STRUCT = "STRUCT"



class ResourceAction:
    id: str
    version: int
    auditData: AuditData
    resource: Resource
    name: str
    title: str
    description: str
    internal: bool
    types: list[SubType]
    input: list[Property]
    output: Property
    annotations: dict[str, str]

    @staticmethod
    def entity_info():
        return EntityInfo("system", "ResourceAction", "system-resourceaction")

File: pkg/resources/resource-action.go
package resources

import (
	"github.com/apibrew/apibrew/pkg/model"
	"github.com/apibrew/apibrew/pkg/resources/special"
	"github.com/apibrew/apibrew/pkg/service/annotations"
	"github.com/apibrew/apibrew/pkg/util"
	"google.golang.org/protobuf/types/known/structpb"
)

var ResourceActionResource = &model.Resource{
	Name:      "ResourceAction",
	Namespace: "system",
	SourceConfig: &model.ResourceSourceConfig{
		DataSource: "system",
		Entity:     "resource_action",
	},
	Types: []*model.ResourceSubType{
		SubTypeType,
		PropertyType,
		special.AuditDataSubType,
	},
	Properties: []*model.ResourceProperty{
		special.IdProperty,
		special.VersionProperty,
		special.AuditProperty,
		{
			Name:     "resource",
			Type:     model.ResourceProperty_REFERENCE,
			Required: true,
			Reference: &model.Reference{
				Resource:  ResourceResource.Name,
				Namespace: ResourceResource.Namespace,
				Cascade:   false,
			},
		},
		{
			Name:     "name",
			Type:     model.ResourceProperty_STRING,
			Length:   256,
			Required: true,
			Unique:   false,
			Annotations: map[string]string{
				annotations.IsHclLabel: annotations.Enabled,
			},
		},
		{
			Name:     "title",
			Type:     model.ResourceProperty_STRING,
			Length:   256,
			Required: false,
			Unique:   false,
			Annotations: map[string]string{
				annotations.IsHclLabel: annotations.Enabled,
			},
		},
		{
			Name:     "description",
			Type:     model.ResourceProperty_STRING,
			Length:   256,
			Required: false,
			Unique:   false,
			Annotations: map[string]string{
				annotations.IsHclLabel: annotations.Enabled,
			},
		},
		{
			Name:         "internal",
			Type:         model.ResourceProperty_BOOL,
			Required:     true,
			DefaultValue: structpb.NewBoolValue(false),
			Annotations: map[string]string{
				annotations.IsHclLabel: annotations.Enabled,
			},
		},
		{
			Name:     "types",
			Type:     model.ResourceProperty_LIST,
			Required: false,
			Item: &model.ResourceProperty{
				Type:    model.ResourceProperty_STRUCT,
				TypeRef: util.Pointer("SubType"),
			},
		},
		{
			Name: "input",
			Type: model.ResourceProperty_MAP,
			Item: &model.ResourceProperty{
				Type:    model.ResourceProperty_STRUCT,
				TypeRef: util.Pointer("Property"),
			},
		},
		{
			Name:    "output",
			Type:    model.ResourceProperty_STRUCT,
			TypeRef: util.Pointer("Property"),
		},
		special.AnnotationsProperty,
	},
	Indexes: []*model.ResourceIndex{
		{
			Properties: []*model.ResourceIndexProperty{
				{
					Name: "resource",
				},
				{
					Name: "name",
				},
			},
			Unique: true,
		},
	},
	Annotations: map[string]string{
		annotations.EnableAudit:     annotations.Enabled,
		annotations.RestApiDisabled: annotations.Enabled,
		annotations.OpenApiGroup:    OpenApiMeta,
	},
}
